client/src/js/directives/findpatient.js

```diff
@@ -11,106 +9,265 @@ FindPatientDirective.$inject = [
+      * @param {string} text Patient name (first_name, middle_name or last_name)
+      *
+      * @description This function make a call to BHIMA API for getting patients
+      * according the name (first_name, middle_name or last_name).
+      *
+      * @return {Array} An array of patients
+      */
+      function searchByName(text) {
+        session.loadStatus = 'loading';
+
+        var url = '/patients/search/?name=';
+        return $http.get(url + text.toLowerCase() + '&limit=20')
         .then(function (response) {
           return response.data;
-        });
+        })
+        .catch(errorHandler)
+        .finally();
       }
 
-      // calls bhima API with uuid
-      function searchByUuid(uuid) {
-        var url = '/patient/';
-        $http.get(url + uuid)
-        .then(function (response) { selectPatient(response.data); })
-        .catch(handler)
-        .finally();
+      /**
+      * @function submit
+      *
+      * @description This function is responsible for call the appropriate function
+      * according we have a search by ID or a search by Name to get data
+      */
+      function submit() {
+        if (session.selected.key === 0 && session.idInput) {
+          searchByReference(session.idInput);
+
+        } else if (session.selected.key === 1 && session.nameInput) {
+          selectPatient(session.nameInput);
+        }
       }
 
-      // make a pretty label
-      function fmtPatient(p) {
-        return p ? p.first_name + ' ' + p.last_name + ' ' + p.middle_name : '';
+      /**
+      * @function findBy
+      *
+      * @param {object} option The selected option
+      *
+      * @description This function is responsible for setting the selected option
+      * between ID or Name option of search
+      */
+      function findBy(option) {
+        session.selected   = option;
+        session.loadStatus = null;
+        session.idInput    = undefined;
+        session.nameInput  = undefined;
+        saveOption(session.selected);
       }
 
-      // change the input type
-      function toggleSearch(s) {
-        session.state = s;
-        saveState({ state : s });
+      /**
+      * @function reload
+      *
+      * @description This function is responsible for enabling the user to input data
+      * again for search by showing the inputs zones (search by ID or by name) again.
+      */
+      function reload() {
+        session.showSearchView = true;
       }
 
-      // expose to view
-      scope.searchReference = searchReference;
-      scope.fuzzyNameSearch = fuzzyNameSearch;
-      scope.toggleSearch = toggleSearch;
-      scope.fmtPatient = fmtPatient;
+      /**
+      * @function formatPatient
+      *
+      * @param {object} patient The patient object
+      *
+      * @description This function is responsible for formatting the patient name
+      * for to be more readable
+      *
+      * @returns {string} The formatted patient name
+      */
+      function formatPatient(p) {
```

Comment:
Instead of calling this function from the view (and adding listeners to the view), how about you call this when the data is loaded from the server?  Something like:
```js
$http.get('/url')
.then(function (response) {
  return response.data;
})
.then(function (patients) {
  patients.forEach(function (patient) {
    patient.label = formatPatient(patient);
  });
  return patients;
});
```
Then you will only format patients when you have new data, instead of on view updates.